### ISSUE TITLE
Add UART functionality to the UPduino

### ISSUE THREAD

**Issue by vr2045** [tinyvision-ai-inc/UPduino-v3.0#3]
The current UPduino 3.0 design doesnt allow the UART to be used if the FPGA is accessing the Flash.
Designs like the RISCV use the Flash but also need access to the UART at the same time.
- Add UART (non flow controlled) access to the UPduino independent of the flash
- Keep cost the same
- Minimize use of extra pins, ideally keep this pin compatible with the UPduino 2.x

**Comment by classic-gentleman** [comment]
Would you mind detailing this issue a bit more? Is it a hardware limitation of some sort, like GPIO sharing, or could it be fixed in the bitstream?

**Comment by vr2045** [comment]
In the current design, the FTDI part doesnt do anything after the FPGA boots up but its UART pins (13,14) are connected to SCK and MOSI respectively which belongs to the flash. This is a HW limitation of the current design. This cannot be fixed by changes in the bitstream...

There are a couple of approaches possible here:
1. Change the FTDI part to be a dual channel FT2232HQ device as used on the iCEbreaker etc.
The FTDI part has a single channel serializer engine, unlike the FT2232 which has 2 such engines. Going to the 2 channel part is possible but significantly increases cost (>$2 BoM) and also increases board area resulting in pretty much full rerouting. As low cost is a primary ask from the community, option though used on projects like the iCE Breaker, is rejected.

2. Add digital muxes to allow the FTDI UART pins to be mux'ed when the SPI is not active.
In this approach, we would add a mux to pins 13, 14 of the FTDI part that would connect these pins to the ICE_SCK/MOSI pins when the SPI is active (SS is low) and otherwise, to a couple of FPGA pins when the SPI is inactive (SS is high).

# Which pins on the FPGA should be used?
Two pins (IOB_18A, IOB_25B_G3) were test points in UPduino v2 and so not normally used by users. This gives us 2 possible pins to use for the UART on the FPGA side.

**Comment by vr2045** [comment]
A possible solution:
![image](https://user-images.githubusercontent.com/29843211/78698312-fe0f0680-78b6-11ea-8e93-abae84d2e354.png)
This is a minimal impact on cost since the 126 device can replace the current 2 gate part used for clock distribution. Also, from the routing point of view, this should be relatively easy.

**Comment by classic-gentleman** [comment]
Might be a good a time as any to revisit the STM32 as a replacement for the FTDI.

**Comment by paulossant** [comment]
Could this repository be a good starting point? https://github.com/ayushgaud/USB_Serial_Bridge

**Comment by vr2045** [comment]
I like the idea of getting rid of the FTDI part, its not only expensive but also limited in functioanlity as compared to an STM32. This board is designed for various FPGA tools like the icestorm/icestudio/apio and Lattice Radiant programmer. All of these tools use the FTDI USB VID/PID signature at present to identify the presence of a valid programmer.

Do you think its kosher to use the VID/PID of the FTDI chip when the USB enumerates in the STM32 part? I think there may be legal trouble here.

My suggestion is to start a different project that uses the STM32 as the basis and push the new USB signature into at least the open source tools and then work with Lattice to enable their programmer to use this board.

Comments?

**Comment by classic-gentleman** [comment]
I have no culinary issues whatsoever with impersonating an USB ID. 🤷‍♂️

In a more serious tone, interface interoperability is an explicit exemption in the copyright law as long as the reverse engineering follows the rules of clean room implementation.

And odds are many other protocols are just as supported, like plain serial comm. It's just a matter of choice. Implementing the full FTDI command set is more trouble than what's worth.

**Comment by vr2045** [comment]
Got it. So what we need for full backward compatibility is a rather high bar and really doesnt add any value. I dont see the point of it except that these boards need to work with the Lattice toolchain, somethign to consider given that this is targeted at a Lattice part.
1. The tinyFPGA did away with this issue by using the Valenty USB stack in the FPGA itself and a dual stage bootloader. This _should_ be supported already in the current 3.0 design but not tested.
2. Another approach that you're suggesting is to use the STM32 as the USB interface. I really support this as the Micro will be usable for a great many things other than just programming.

**Comment by classic-gentleman** [comment]
I never used Lattice's tools, I'm 100% invested in the open source toolchains. So it's a matter of figuring out the intersection of what both support and if more than a single option emerges, decide what's the best.

I don't see a single case where using STM32 as the USB interface isn't a win, and having it as a coprocessor for whatever is super welcome as well.

**Comment by vr2045** [comment]
I agree with you completely @classic-gentleman. However, I dont think this belongs in the UPduino which is designed to work with whatever tools commercial/open source are out there.
If you dont mind, we can start a new Github project where we can build this from the ground up without any baggage.

Here are some related boards that do something similar:
http://ebrombaugh.studionebula.com/embedded/f303_ice5/index.html
https://github.com/tvelliott/dsp_ice
https://folknologylabs.wordpress.com/2016/08/17/the-lull-before-the-storm/

**Comment by classic-gentleman** [comment]
I'm not seeing a conflict there. Am I missing something? Because as far as I can tell it's a matter of flashing the appropriate image on the STM32 so it behaves like anything we want it to be. It's not like we're missing any flexibility going this route.

**Comment by vr2045** [comment]
We're in agreement that a thing with an STM32 + iCE40 would make sense. I think where we differ may be in how to move forward.

I see the requirement to be fully backwards compatible if we stick with the UPduino name. This has been codified in many places that we dont have control over as being FTDI (USB VIP/PID) + MPSSE engine using specific FTDI routines. eg. iceprog uses libftdi, Lattice tools use the same and so on.

This means we're bogged down trying to ensure everyone who uses the tools today are kept happy which means re-implementing the FTDI commands (which I think we agree is a waste of time). In order to get to a stable backward compatible state, a tremendous amount of testing, debug and support is going to be required, resulting in an FTDI clone (of course, with other capabilities) which is not the main aspect of the project. IMHO, this would be the first STM SW deliverable in order to say we have something thats an UPduino.

Wouldnt it be faster to develop if we called it something else (ARMistiCE? :) ) so we can start from scratch and use whatever protocol makes the best sense to communicate with the host machine & program FPGA's (drag/drop, command line using UART/USB etc.)?

This device could implement the FTDI protocol at some point in the future to take advantage of the current ecosystem if that makes sense to do (I'm not convinced this is required BTW).

**Comment by classic-gentleman** [comment]
The name alone had me sold. For the love of _deity_ let's register that organisation and feel free to add me there!

**Comment by vr2045** [comment]
:) I created a project with that name and added you with write access.
My suggestion is to agree on clear goals and then move forward...

**Comment by paulossant** [comment]
Can i provide some help? I know stm32 toolchain and now starting with fpga boards.

**Comment by vr2045** [comment]
Sure, that would be great! I started a wiki for the project so we can put our thoughts down, agree on it and move forward. I added you as a contributor as well.
I can create a discord channel for this on my [discord server](https://discord.gg/hJ98Zbk) as well so we can have real time discussions instead of this sort of delayed github chatter.

**Comment by classic-gentleman** [comment]
> Can i provide some help? I know stm32 toolchain and now starting with fpga boards.

Any helping hand is welcome! I'd be especially interested in loading bitstreams via UF2. Something for interfacing with the FPGA for debugging would be nice as well, but I'm not experienced in much beyond OpenOCD...

**Comment by vr2045** [comment]
Any comments on the attached schematic to allow the UART functionality that this thread was originally about to work?
## SPI operation
Whenever the SSN line goes active (low), it will enable the normal signal route.

## UART operation
When the SSN line is inactive (high), the FTDI SCK(UART_TXD) and MOSI (UART_RXD) get hooked up to alternate pins on the FPGA and the SPI-Flash signals to the FPGA/Flash get isolated so they dont combine with the UART.
[UPduino_v3.0.pdf](https://github.com/tinyvision-ai-inc/UPduino-v3.0/files/4451670/UPduino_v3.0.pdf)

**Comment by classic-gentleman** [comment]
To be honest I didn't look at the schematic in detail because I don't have a clear idea about what the issue is. SPI is supposed to be a shared bus. The slave devices are supposed to high-Z the lines when CS is not asserted. Unless we're dealing with bad citizens I fail to understand the issue.

But as mentioned I didn't take a look there. I'm going by theory. Might be the async nature of UART if there's no buffer/interrupt to signal activity back to the master.

**Comment by vr2045** [comment]
OK, let me restate this, hopefully more clearly!
The UPduino can be connected to a USB for SPI and/or UART connectivity as long as the Flash pins are used for this functionality.
In situations where the flash is being accessed by the design (such as a RISCV running from flash), the UART functionality cannot be used and requires the user to connect an external UART-USB.

This is typically solved using a second channel on the FTDI. This is an expensive solution.

An alternative is to add a couple of logic gates to have this functionality which is what the schematic changes do...

**Comment by bsiever** [comment]
This is a very old thread, but a solution that fits some needs with the existing hardware:
1. Boot [start.s](https://github.com/digital-logic-and-computer-design/picorv32/blob/cse260m/picosoc/start.s) from flash and have a simple bootloader process that transfers everything else (including code/constants) to RAM.  This requires an appropriately configured [linker script](https://github.com/digital-logic-and-computer-design/picorv32/blob/cse260m/picosoc/sections.lds). 
2. The bootloader pivots to execution of a small [startup routine](https://github.com/digital-logic-and-computer-design/picorv32/blob/016e8d7b89135d5be9734b8c44aaebb9725f4e7c/picosoc/start.s#L167) in RAM, which disables the flash access, enables UART, and starts the user application (i.e., calls `main`).

All links above are for a specific configuration and provided a proof-of-concept.  They may need to be adapted for other uses.

**Comment by vr2045** [comment]
Thats a great hack to get this to work with no HW changes.
The limitation is that this works for programs that fit completely into RAM. Good enough for small projects.
